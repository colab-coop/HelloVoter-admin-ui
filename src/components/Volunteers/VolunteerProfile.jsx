import React, { useState } from "react";
import Button from "@material-ui/core/Button";
import { notify_error } from "../../common.js";

const VALIDATION_ERROR_MESSAGE = "All fields are required.";

const _style = {
  profile: {
    width: "30%",
  },
  field: {
    width: "100%",
  },
  invalidField: {
    width: "100%",
    border: "1px solid red",
    backgroundColor: "rgba(255, 0, 0, 0.2)",
  },
};

export const ProfileField = ({
  editing,
  type,
  label,
  field,
  value,
  onChange,
  invalidFields,
}) =>
  editing ? (
    <div>
      {label}:{" "}
      <input
        style={invalidFields[field] ? _style.invalidField : _style.field}
        name={field}
        type={type}
        onChange={onChange}
        value={value || ""}
      />
    </div>
  ) : (
    <div>
      {label}: {value}
    </div>
  );

const extractFormData = (volunteer) => ({
  first_name: volunteer.first_name,
  last_name: volunteer.last_name,
  date_of_birth: volunteer.date_of_birth,
  address: volunteer.address
    ? {
        address1: volunteer.address.address1,
        address2: volunteer.address.address2,
        city: volunteer.address.city,
        state: volunteer.address.state,
        zip: volunteer.address.zip,
      }
    : volunteer.address,
});

export default ({ volunteer, refer }) => {
  const [editing, setEditing] = useState(false);
  const [formData, setFormData] = useState(extractFormData(volunteer));
  const [invalidFields, setInvalidFields] = useState({});

  const setVolunteerField = (evt) =>
    setFormData({
      ...formData,
      [evt.target.name]: evt.target.value,
    });

  const setAddressField = (evt) =>
    setFormData({
      ...formData,
      address: {
        ...formData.address,
        [evt.target.name]: evt.target.value,
      },
    });

  const save = async () => {
    // validate that all fields are filled (except address line 2)
    const blankFields = {};
    Object.keys(formData).forEach((k) => {
      if (!formData[k]) {
        blankFields[k] = true;
      }
    });
    Object.keys(formData.address).forEach((k) => {
      if (k != "address2" && !formData.address[k]) {
        blankFields[k] = true;
      }
    });
    if (Object.keys(blankFields).length > 0) {
      setInvalidFields(blankFields);
      notify_error(
        new Error(VALIDATION_ERROR_MESSAGE),
        VALIDATION_ERROR_MESSAGE
      );
      return;
    }
    refer._updateAmbassador(volunteer.id, formData);
  };

  const cancel = () => {
    setFormData(extractFormData(volunteer));
    setInvalidFields({});
    setEditing(false);
  };

  const addressData = formData.address || {};

  return (
    <div style={_style.profile}>
      <ProfileField
        editing={editing}
        label="First Name"
        type="text"
        field="first_name"
        value={formData.first_name}
        onChange={setVolunteerField}
        invalidFields={invalidFields}
      />
      <ProfileField
        editing={editing}
        label="Last Name"
        type="text"
        field="last_name"
        value={formData.last_name}
        onChange={setVolunteerField}
        invalidFields={invalidFields}
      />
      <ProfileField
        editing={editing}
        label="Address Line 1"
        type="text"
        field="address1"
        value={addressData.address1}
        onChange={setAddressField}
        invalidFields={invalidFields}
      />
      <ProfileField
        editing={editing}
        label="Address Line 2"
        type="text"
        field="address2"
        value={addressData.address2}
        onChange={setAddressField}
        invalidFields={invalidFields}
      />
      <ProfileField
        editing={editing}
        label="City"
        type="text"
        field="city"
        value={addressData.city}
        onChange={setAddressField}
        invalidFields={invalidFields}
      />
      <ProfileField
        editing={editing}
        label="State"
        type="text"
        field="state"
        value={addressData.state}
        onChange={setAddressField}
        invalidFields={invalidFields}
      />
      <ProfileField
        editing={editing}
        label="Zip"
        type="number"
        field="zip"
        value={addressData.zip}
        onChange={setAddressField}
        invalidFields={invalidFields}
      />
      <ProfileField editing={false} label="Phone" value={volunteer.phone} />
      <ProfileField editing={false} label="Email" value={volunteer.email} />
      <ProfileField
        editing={editing}
        label="Date of Birth"
        type="date"
        field="date_of_birth"
        value={formData.date_of_birth}
        onChange={setVolunteerField}
        invalidFields={invalidFields}
      />
      {editing ? (
        <div>
          <Button onClick={save}>Save</Button>
          <Button onClick={cancel}>Cancel</Button>
        </div>
      ) : (
        <div>
          <Button onClick={() => setEditing(true)}>Edit Profile</Button>
        </div>
      )}
    </div>
  );
};
