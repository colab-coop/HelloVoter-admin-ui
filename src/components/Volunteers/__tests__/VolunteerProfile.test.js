import React from "react";
import { mount } from "enzyme";
import toJson from "enzyme-to-json";
import VolunteerProfile, { ProfileField } from "../VolunteerProfile.jsx";
import Button from "@material-ui/core/Button";

jest.mock("../../../common.js");
import { notify_error } from "../../../common.js";

const VOLUNTEER = {
  id: 1,
  first_name: "Foo",
  last_name: "Bar",
  email: "foo@bar.baz",
  date_of_birth: Date.parse("01-01-1970"),
  address: {
    address1: "123 Test St",
    address2: "Apt 4",
    city: "Test City",
    state: "GA",
    zip: "00000",
  },
};

const VOLUNTEER_UPDATE = {
  first_name: "Updated",
  last_name: "Volunteer",
  date_of_birth: Date.parse("01-02-1970"),
  address: {
    address1: "123 Test Rd",
    city: "Updated City",
    state: "FL",
    zip: "00001",
  },
};

const READ_ONLY_FIELDS = ["Phone", "Email"];

describe("VolunteerProfile", () => {
  it("should match snapshot", () => {
    const refer = { _updateAmbassador: jest.fn() };
    const profile = mount(
      <VolunteerProfile volunteer={VOLUNTEER} refer={refer} />
    );
    expect(toJson(profile)).toMatchSnapshot();
  });

  it("should match snapshot of editing state", () => {
    const refer = { _updateAmbassador: jest.fn() };
    const profile = mount(
      <VolunteerProfile volunteer={VOLUNTEER} refer={refer} />
    );
    profile.find(Button).simulate("click");
    expect(toJson(profile)).toMatchSnapshot();
  });

  it("switches to editing state when the edit profile button is pressed", () => {
    const refer = { _updateAmbassador: jest.fn() };
    const profile = mount(
      <VolunteerProfile volunteer={VOLUNTEER} refer={refer} />
    );
    profile.find(ProfileField).forEach((f) => {
      expect(f.prop("editing")).toBeFalsy();
      expect(f.find("input").length).toBe(0);
    });
    let buttons = profile.find(Button);
    expect(buttons.text()).toBe("Edit Profile");
    buttons.simulate("click");
    profile.find(ProfileField).forEach((f) => {
      if (!READ_ONLY_FIELDS.includes(f.prop("label"))) {
        expect(f.prop("editing")).toBeTruthy();
        expect(f.find("input").length).toBe(1);
      }
    });
    buttons = profile.find(Button);
    expect(buttons.length).toBe(2);
    expect(buttons.at(0).text()).toBe("Save");
    expect(buttons.at(1).text()).toBe("Cancel");
  });

  it("exits editing state when the cancel button is pressed", () => {
    const refer = { _updateAmbassador: jest.fn() };
    const profile = mount(
      <VolunteerProfile volunteer={VOLUNTEER} refer={refer} />
    );
    let buttons = profile.find(Button);
    buttons.simulate("click");
    buttons = profile.find(Button);
    buttons.at(1).simulate("click");
    profile.find(ProfileField).forEach((f) => {
      expect(f.prop("editing")).toBeFalsy();
      expect(f.find("input").length).toBe(0);
    });
    buttons = profile.find(Button);
    expect(buttons.text()).toBe("Edit Profile");
  });

  it("saves changes when the save button is pressed", () => {
    const refer = { _updateAmbassador: jest.fn() };
    const profile = mount(
      <VolunteerProfile volunteer={VOLUNTEER} refer={refer} />
    );
    let buttons = profile.find(Button);
    buttons.simulate("click");
    profile.find(ProfileField).forEach((f) => {
      if (!READ_ONLY_FIELDS.includes(f.prop("label"))) {
        const name = f.prop("field");
        const value = VOLUNTEER_UPDATE[name] || VOLUNTEER_UPDATE.address[name];
        f.find("input").simulate("change", { target: { name, value } });
      }
    });
    buttons = profile.find(Button);
    buttons.at(0).simulate("click");
    expect(refer._updateAmbassador.mock.calls.length).toBe(1);
    expect(refer._updateAmbassador.mock.calls[0][0]).toBe(VOLUNTEER.id);
    expect(refer._updateAmbassador.mock.calls[0][1]).toEqual(VOLUNTEER_UPDATE);
  });

  it("rejects inputs if fields other than address2 are left blank", () => {
    const refer = { _updateAmbassador: jest.fn() };
    const profile = mount(
      <VolunteerProfile volunteer={VOLUNTEER} refer={refer} />
    );
    let buttons = profile.find(Button);
    buttons.simulate("click");
    profile.find(ProfileField).forEach((f) => {
      if (!READ_ONLY_FIELDS.includes(f.prop("label"))) {
        const name = f.prop("field");
        const value = undefined;
        f.find("input").simulate("change", { target: { name, value } });
      }
    });
    buttons = profile.find(Button);
    buttons.at(0).simulate("click");
    expect(notify_error.mock.calls.length).toBe(1);
    profile.find(ProfileField).forEach((f) => {
      if (
        f.prop("field") !== "address2" &&
        !READ_ONLY_FIELDS.includes(f.prop("label"))
      ) {
        const inputStyle = f.find("input").prop("style");
        expect(inputStyle).toHaveProperty("border");
        expect(inputStyle).toHaveProperty("backgroundColor");
      }
    });
  });
});
