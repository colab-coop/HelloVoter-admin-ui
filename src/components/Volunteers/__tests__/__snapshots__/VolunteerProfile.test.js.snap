// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`VolunteerProfile should match snapshot 1`] = `
<_default
  refer={
    Object {
      "_updateAmbassador": [MockFunction],
    }
  }
  volunteer={
    Object {
      "address": Object {
        "address1": "123 Test St",
        "address2": "Apt 4",
        "city": "Test City",
        "state": "GA",
        "zip": "00000",
      },
      "date_of_birth": 18000000,
      "email": "foo@bar.baz",
      "first_name": "Foo",
      "id": 1,
      "last_name": "Bar",
    }
  }
>
  <div
    style={
      Object {
        "width": "30%",
      }
    }
  >
    <ProfileField
      editing={false}
      field="first_name"
      invalidFields={Object {}}
      label="First Name"
      onChange={[Function]}
      type="text"
      value="Foo"
    >
      <div>
        First Name
        : 
        Foo
      </div>
    </ProfileField>
    <ProfileField
      editing={false}
      field="last_name"
      invalidFields={Object {}}
      label="Last Name"
      onChange={[Function]}
      type="text"
      value="Bar"
    >
      <div>
        Last Name
        : 
        Bar
      </div>
    </ProfileField>
    <ProfileField
      editing={false}
      field="address1"
      invalidFields={Object {}}
      label="Address Line 1"
      onChange={[Function]}
      type="text"
      value="123 Test St"
    >
      <div>
        Address Line 1
        : 
        123 Test St
      </div>
    </ProfileField>
    <ProfileField
      editing={false}
      field="address2"
      invalidFields={Object {}}
      label="Address Line 2"
      onChange={[Function]}
      type="text"
      value="Apt 4"
    >
      <div>
        Address Line 2
        : 
        Apt 4
      </div>
    </ProfileField>
    <ProfileField
      editing={false}
      field="city"
      invalidFields={Object {}}
      label="City"
      onChange={[Function]}
      type="text"
      value="Test City"
    >
      <div>
        City
        : 
        Test City
      </div>
    </ProfileField>
    <ProfileField
      editing={false}
      field="state"
      invalidFields={Object {}}
      label="State"
      onChange={[Function]}
      type="text"
      value="GA"
    >
      <div>
        State
        : 
        GA
      </div>
    </ProfileField>
    <ProfileField
      editing={false}
      field="zip"
      invalidFields={Object {}}
      label="Zip"
      onChange={[Function]}
      type="number"
      value="00000"
    >
      <div>
        Zip
        : 
        00000
      </div>
    </ProfileField>
    <ProfileField
      editing={false}
      label="Phone"
    >
      <div>
        Phone
        : 
      </div>
    </ProfileField>
    <ProfileField
      editing={false}
      field="email"
      invalidFields={Object {}}
      label="Email"
      onChange={[Function]}
      type="email"
      value="foo@bar.baz"
    >
      <div>
        Email
        : 
        foo@bar.baz
      </div>
    </ProfileField>
    <ProfileField
      editing={false}
      field="date_of_birth"
      invalidFields={Object {}}
      label="Date of Birth"
      onChange={[Function]}
      type="date"
      value={18000000}
    >
      <div>
        Date of Birth
        : 
        18000000
      </div>
    </ProfileField>
    <div>
      <WithStyles(ForwardRef(Button))
        onClick={[Function]}
      >
        <ForwardRef(Button)
          classes={
            Object {
              "colorInherit": "MuiButton-colorInherit",
              "contained": "MuiButton-contained",
              "containedPrimary": "MuiButton-containedPrimary",
              "containedSecondary": "MuiButton-containedSecondary",
              "containedSizeLarge": "MuiButton-containedSizeLarge",
              "containedSizeSmall": "MuiButton-containedSizeSmall",
              "disableElevation": "MuiButton-disableElevation",
              "disabled": "Mui-disabled",
              "endIcon": "MuiButton-endIcon",
              "focusVisible": "Mui-focusVisible",
              "fullWidth": "MuiButton-fullWidth",
              "iconSizeLarge": "MuiButton-iconSizeLarge",
              "iconSizeMedium": "MuiButton-iconSizeMedium",
              "iconSizeSmall": "MuiButton-iconSizeSmall",
              "label": "MuiButton-label",
              "outlined": "MuiButton-outlined",
              "outlinedPrimary": "MuiButton-outlinedPrimary",
              "outlinedSecondary": "MuiButton-outlinedSecondary",
              "outlinedSizeLarge": "MuiButton-outlinedSizeLarge",
              "outlinedSizeSmall": "MuiButton-outlinedSizeSmall",
              "root": "MuiButton-root",
              "sizeLarge": "MuiButton-sizeLarge",
              "sizeSmall": "MuiButton-sizeSmall",
              "startIcon": "MuiButton-startIcon",
              "text": "MuiButton-text",
              "textPrimary": "MuiButton-textPrimary",
              "textSecondary": "MuiButton-textSecondary",
              "textSizeLarge": "MuiButton-textSizeLarge",
              "textSizeSmall": "MuiButton-textSizeSmall",
            }
          }
          onClick={[Function]}
        >
          <WithStyles(ForwardRef(ButtonBase))
            className="MuiButton-root MuiButton-text"
            component="button"
            disabled={false}
            focusRipple={true}
            focusVisibleClassName="Mui-focusVisible"
            onClick={[Function]}
            type="button"
          >
            <ForwardRef(ButtonBase)
              className="MuiButton-root MuiButton-text"
              classes={
                Object {
                  "disabled": "Mui-disabled",
                  "focusVisible": "Mui-focusVisible",
                  "root": "MuiButtonBase-root",
                }
              }
              component="button"
              disabled={false}
              focusRipple={true}
              focusVisibleClassName="Mui-focusVisible"
              onClick={[Function]}
              type="button"
            >
              <button
                className="MuiButtonBase-root MuiButton-root MuiButton-text"
                disabled={false}
                onBlur={[Function]}
                onClick={[Function]}
                onDragLeave={[Function]}
                onFocus={[Function]}
                onKeyDown={[Function]}
                onKeyUp={[Function]}
                onMouseDown={[Function]}
                onMouseLeave={[Function]}
                onMouseUp={[Function]}
                onTouchEnd={[Function]}
                onTouchMove={[Function]}
                onTouchStart={[Function]}
                tabIndex={0}
                type="button"
              >
                <span
                  className="MuiButton-label"
                >
                  Edit Profile
                </span>
                <NoSsr>
                  <WithStyles(memo)
                    center={false}
                  >
                    <ForwardRef(TouchRipple)
                      center={false}
                      classes={
                        Object {
                          "child": "MuiTouchRipple-child",
                          "childLeaving": "MuiTouchRipple-childLeaving",
                          "childPulsate": "MuiTouchRipple-childPulsate",
                          "ripple": "MuiTouchRipple-ripple",
                          "ripplePulsate": "MuiTouchRipple-ripplePulsate",
                          "rippleVisible": "MuiTouchRipple-rippleVisible",
                          "root": "MuiTouchRipple-root",
                        }
                      }
                    >
                      <span
                        className="MuiTouchRipple-root"
                      >
                        <TransitionGroup
                          childFactory={[Function]}
                          component={null}
                          exit={true}
                        />
                      </span>
                    </ForwardRef(TouchRipple)>
                  </WithStyles(memo)>
                </NoSsr>
              </button>
            </ForwardRef(ButtonBase)>
          </WithStyles(ForwardRef(ButtonBase))>
        </ForwardRef(Button)>
      </WithStyles(ForwardRef(Button))>
    </div>
  </div>
</_default>
`;

exports[`VolunteerProfile should match snapshot of editing state 1`] = `
<_default
  refer={
    Object {
      "_updateAmbassador": [MockFunction],
    }
  }
  volunteer={
    Object {
      "address": Object {
        "address1": "123 Test St",
        "address2": "Apt 4",
        "city": "Test City",
        "state": "GA",
        "zip": "00000",
      },
      "date_of_birth": 18000000,
      "email": "foo@bar.baz",
      "first_name": "Foo",
      "id": 1,
      "last_name": "Bar",
    }
  }
>
  <div
    style={
      Object {
        "width": "30%",
      }
    }
  >
    <ProfileField
      editing={true}
      field="first_name"
      invalidFields={Object {}}
      label="First Name"
      onChange={[Function]}
      type="text"
      value="Foo"
    >
      <div>
        First Name
        :
         
        <input
          name="first_name"
          onChange={[Function]}
          style={
            Object {
              "width": "100%",
            }
          }
          type="text"
          value="Foo"
        />
      </div>
    </ProfileField>
    <ProfileField
      editing={true}
      field="last_name"
      invalidFields={Object {}}
      label="Last Name"
      onChange={[Function]}
      type="text"
      value="Bar"
    >
      <div>
        Last Name
        :
         
        <input
          name="last_name"
          onChange={[Function]}
          style={
            Object {
              "width": "100%",
            }
          }
          type="text"
          value="Bar"
        />
      </div>
    </ProfileField>
    <ProfileField
      editing={true}
      field="address1"
      invalidFields={Object {}}
      label="Address Line 1"
      onChange={[Function]}
      type="text"
      value="123 Test St"
    >
      <div>
        Address Line 1
        :
         
        <input
          name="address1"
          onChange={[Function]}
          style={
            Object {
              "width": "100%",
            }
          }
          type="text"
          value="123 Test St"
        />
      </div>
    </ProfileField>
    <ProfileField
      editing={true}
      field="address2"
      invalidFields={Object {}}
      label="Address Line 2"
      onChange={[Function]}
      type="text"
      value="Apt 4"
    >
      <div>
        Address Line 2
        :
         
        <input
          name="address2"
          onChange={[Function]}
          style={
            Object {
              "width": "100%",
            }
          }
          type="text"
          value="Apt 4"
        />
      </div>
    </ProfileField>
    <ProfileField
      editing={true}
      field="city"
      invalidFields={Object {}}
      label="City"
      onChange={[Function]}
      type="text"
      value="Test City"
    >
      <div>
        City
        :
         
        <input
          name="city"
          onChange={[Function]}
          style={
            Object {
              "width": "100%",
            }
          }
          type="text"
          value="Test City"
        />
      </div>
    </ProfileField>
    <ProfileField
      editing={true}
      field="state"
      invalidFields={Object {}}
      label="State"
      onChange={[Function]}
      type="text"
      value="GA"
    >
      <div>
        State
        :
         
        <input
          name="state"
          onChange={[Function]}
          style={
            Object {
              "width": "100%",
            }
          }
          type="text"
          value="GA"
        />
      </div>
    </ProfileField>
    <ProfileField
      editing={true}
      field="zip"
      invalidFields={Object {}}
      label="Zip"
      onChange={[Function]}
      type="number"
      value="00000"
    >
      <div>
        Zip
        :
         
        <input
          name="zip"
          onChange={[Function]}
          style={
            Object {
              "width": "100%",
            }
          }
          type="number"
          value="00000"
        />
      </div>
    </ProfileField>
    <ProfileField
      editing={false}
      label="Phone"
    >
      <div>
        Phone
        : 
      </div>
    </ProfileField>
    <ProfileField
      editing={true}
      field="email"
      invalidFields={Object {}}
      label="Email"
      onChange={[Function]}
      type="email"
      value="foo@bar.baz"
    >
      <div>
        Email
        :
         
        <input
          name="email"
          onChange={[Function]}
          style={
            Object {
              "width": "100%",
            }
          }
          type="email"
          value="foo@bar.baz"
        />
      </div>
    </ProfileField>
    <ProfileField
      editing={true}
      field="date_of_birth"
      invalidFields={Object {}}
      label="Date of Birth"
      onChange={[Function]}
      type="date"
      value={18000000}
    >
      <div>
        Date of Birth
        :
         
        <input
          name="date_of_birth"
          onChange={[Function]}
          style={
            Object {
              "width": "100%",
            }
          }
          type="date"
          value={18000000}
        />
      </div>
    </ProfileField>
    <div>
      <WithStyles(ForwardRef(Button))
        onClick={[Function]}
      >
        <ForwardRef(Button)
          classes={
            Object {
              "colorInherit": "MuiButton-colorInherit",
              "contained": "MuiButton-contained",
              "containedPrimary": "MuiButton-containedPrimary",
              "containedSecondary": "MuiButton-containedSecondary",
              "containedSizeLarge": "MuiButton-containedSizeLarge",
              "containedSizeSmall": "MuiButton-containedSizeSmall",
              "disableElevation": "MuiButton-disableElevation",
              "disabled": "Mui-disabled",
              "endIcon": "MuiButton-endIcon",
              "focusVisible": "Mui-focusVisible",
              "fullWidth": "MuiButton-fullWidth",
              "iconSizeLarge": "MuiButton-iconSizeLarge",
              "iconSizeMedium": "MuiButton-iconSizeMedium",
              "iconSizeSmall": "MuiButton-iconSizeSmall",
              "label": "MuiButton-label",
              "outlined": "MuiButton-outlined",
              "outlinedPrimary": "MuiButton-outlinedPrimary",
              "outlinedSecondary": "MuiButton-outlinedSecondary",
              "outlinedSizeLarge": "MuiButton-outlinedSizeLarge",
              "outlinedSizeSmall": "MuiButton-outlinedSizeSmall",
              "root": "MuiButton-root",
              "sizeLarge": "MuiButton-sizeLarge",
              "sizeSmall": "MuiButton-sizeSmall",
              "startIcon": "MuiButton-startIcon",
              "text": "MuiButton-text",
              "textPrimary": "MuiButton-textPrimary",
              "textSecondary": "MuiButton-textSecondary",
              "textSizeLarge": "MuiButton-textSizeLarge",
              "textSizeSmall": "MuiButton-textSizeSmall",
            }
          }
          onClick={[Function]}
        >
          <WithStyles(ForwardRef(ButtonBase))
            className="MuiButton-root MuiButton-text"
            component="button"
            disabled={false}
            focusRipple={true}
            focusVisibleClassName="Mui-focusVisible"
            onClick={[Function]}
            type="button"
          >
            <ForwardRef(ButtonBase)
              className="MuiButton-root MuiButton-text"
              classes={
                Object {
                  "disabled": "Mui-disabled",
                  "focusVisible": "Mui-focusVisible",
                  "root": "MuiButtonBase-root",
                }
              }
              component="button"
              disabled={false}
              focusRipple={true}
              focusVisibleClassName="Mui-focusVisible"
              onClick={[Function]}
              type="button"
            >
              <button
                className="MuiButtonBase-root MuiButton-root MuiButton-text"
                disabled={false}
                onBlur={[Function]}
                onClick={[Function]}
                onDragLeave={[Function]}
                onFocus={[Function]}
                onKeyDown={[Function]}
                onKeyUp={[Function]}
                onMouseDown={[Function]}
                onMouseLeave={[Function]}
                onMouseUp={[Function]}
                onTouchEnd={[Function]}
                onTouchMove={[Function]}
                onTouchStart={[Function]}
                tabIndex={0}
                type="button"
              >
                <span
                  className="MuiButton-label"
                >
                  Save
                </span>
                <NoSsr>
                  <WithStyles(memo)
                    center={false}
                  >
                    <ForwardRef(TouchRipple)
                      center={false}
                      classes={
                        Object {
                          "child": "MuiTouchRipple-child",
                          "childLeaving": "MuiTouchRipple-childLeaving",
                          "childPulsate": "MuiTouchRipple-childPulsate",
                          "ripple": "MuiTouchRipple-ripple",
                          "ripplePulsate": "MuiTouchRipple-ripplePulsate",
                          "rippleVisible": "MuiTouchRipple-rippleVisible",
                          "root": "MuiTouchRipple-root",
                        }
                      }
                    >
                      <span
                        className="MuiTouchRipple-root"
                      >
                        <TransitionGroup
                          childFactory={[Function]}
                          component={null}
                          exit={true}
                        />
                      </span>
                    </ForwardRef(TouchRipple)>
                  </WithStyles(memo)>
                </NoSsr>
              </button>
            </ForwardRef(ButtonBase)>
          </WithStyles(ForwardRef(ButtonBase))>
        </ForwardRef(Button)>
      </WithStyles(ForwardRef(Button))>
      <WithStyles(ForwardRef(Button))
        onClick={[Function]}
      >
        <ForwardRef(Button)
          classes={
            Object {
              "colorInherit": "MuiButton-colorInherit",
              "contained": "MuiButton-contained",
              "containedPrimary": "MuiButton-containedPrimary",
              "containedSecondary": "MuiButton-containedSecondary",
              "containedSizeLarge": "MuiButton-containedSizeLarge",
              "containedSizeSmall": "MuiButton-containedSizeSmall",
              "disableElevation": "MuiButton-disableElevation",
              "disabled": "Mui-disabled",
              "endIcon": "MuiButton-endIcon",
              "focusVisible": "Mui-focusVisible",
              "fullWidth": "MuiButton-fullWidth",
              "iconSizeLarge": "MuiButton-iconSizeLarge",
              "iconSizeMedium": "MuiButton-iconSizeMedium",
              "iconSizeSmall": "MuiButton-iconSizeSmall",
              "label": "MuiButton-label",
              "outlined": "MuiButton-outlined",
              "outlinedPrimary": "MuiButton-outlinedPrimary",
              "outlinedSecondary": "MuiButton-outlinedSecondary",
              "outlinedSizeLarge": "MuiButton-outlinedSizeLarge",
              "outlinedSizeSmall": "MuiButton-outlinedSizeSmall",
              "root": "MuiButton-root",
              "sizeLarge": "MuiButton-sizeLarge",
              "sizeSmall": "MuiButton-sizeSmall",
              "startIcon": "MuiButton-startIcon",
              "text": "MuiButton-text",
              "textPrimary": "MuiButton-textPrimary",
              "textSecondary": "MuiButton-textSecondary",
              "textSizeLarge": "MuiButton-textSizeLarge",
              "textSizeSmall": "MuiButton-textSizeSmall",
            }
          }
          onClick={[Function]}
        >
          <WithStyles(ForwardRef(ButtonBase))
            className="MuiButton-root MuiButton-text"
            component="button"
            disabled={false}
            focusRipple={true}
            focusVisibleClassName="Mui-focusVisible"
            onClick={[Function]}
            type="button"
          >
            <ForwardRef(ButtonBase)
              className="MuiButton-root MuiButton-text"
              classes={
                Object {
                  "disabled": "Mui-disabled",
                  "focusVisible": "Mui-focusVisible",
                  "root": "MuiButtonBase-root",
                }
              }
              component="button"
              disabled={false}
              focusRipple={true}
              focusVisibleClassName="Mui-focusVisible"
              onClick={[Function]}
              type="button"
            >
              <button
                className="MuiButtonBase-root MuiButton-root MuiButton-text"
                disabled={false}
                onBlur={[Function]}
                onClick={[Function]}
                onDragLeave={[Function]}
                onFocus={[Function]}
                onKeyDown={[Function]}
                onKeyUp={[Function]}
                onMouseDown={[Function]}
                onMouseLeave={[Function]}
                onMouseUp={[Function]}
                onTouchEnd={[Function]}
                onTouchMove={[Function]}
                onTouchStart={[Function]}
                tabIndex={0}
                type="button"
              >
                <span
                  className="MuiButton-label"
                >
                  Cancel
                </span>
                <NoSsr>
                  <WithStyles(memo)
                    center={false}
                  >
                    <ForwardRef(TouchRipple)
                      center={false}
                      classes={
                        Object {
                          "child": "MuiTouchRipple-child",
                          "childLeaving": "MuiTouchRipple-childLeaving",
                          "childPulsate": "MuiTouchRipple-childPulsate",
                          "ripple": "MuiTouchRipple-ripple",
                          "ripplePulsate": "MuiTouchRipple-ripplePulsate",
                          "rippleVisible": "MuiTouchRipple-rippleVisible",
                          "root": "MuiTouchRipple-root",
                        }
                      }
                    >
                      <span
                        className="MuiTouchRipple-root"
                      >
                        <TransitionGroup
                          childFactory={[Function]}
                          component={null}
                          exit={true}
                        />
                      </span>
                    </ForwardRef(TouchRipple)>
                  </WithStyles(memo)>
                </NoSsr>
              </button>
            </ForwardRef(ButtonBase)>
          </WithStyles(ForwardRef(ButtonBase))>
        </ForwardRef(Button)>
      </WithStyles(ForwardRef(Button))>
    </div>
  </div>
</_default>
`;
